block/src/Pos/Block/Network/Types.hs

```diff
@@ -66,41 +75,111 @@ instance Buildable MsgGetBlocks where
+        case tag of
+            0 -> MsgHeaders <$> decode
+            1 -> MsgNoHeaders <$> decode
+            t -> cborError $ "MsgHeaders wrong tag: " <> show t
+
 -- | 'Block' message (see protocol specification).
 data MsgBlock
     = MsgBlock Block
     | MsgNoBlock Text
     deriving (Eq, Show, Generic)
 
+instance Bi MsgBlock where
+    encode = \case
+        MsgBlock b -> encodeListLen 2 <> encode (0 :: Word8) <> encode b
+        MsgNoBlock t -> encodeListLen 2 <> encode (1 :: Word8) <> encode t
+    decode = do
+        enforceSize "MsgBlock" 2
+        tag <- decode @Word8
+        case tag of
+            0 -> MsgBlock <$> decode
+            1 -> MsgNoBlock <$> decode
+            t -> cborError $ "MsgBlock wrong tag: " <> show t
+
 -- | 'SerializedBlock' message
 data MsgSerializedBlock
     = MsgSerializedBlock SerializedBlock
     | MsgNoSerializedBlock Text
     deriving (Generic)
 
-data MsgStream
-    = MsgStart MsgStreamStart
-    | MsgUpdate MsgStreamUpdate
-    deriving (Eq, Show, Generic)
-
 data MsgStreamStart = MsgStreamStart
     { mssFrom   :: ![HeaderHash]
     , mssTo     :: !HeaderHash
     , mssWindow :: !Word32
     } deriving (Generic, Show, Eq)
 
+deriveSimpleBi ''MsgStreamStart [
+    Cons 'MsgStreamStart [
+        Field [| mssFrom   :: [HeaderHash] |],
+        Field [| mssTo     :: HeaderHash |],
+        Field [| mssWindow :: Word32 |]
+    ]]
+
 data MsgStreamUpdate = MsgStreamUpdate
     { msuWindow :: !Word32
     } deriving (Generic, Show, Eq)
 
+deriveSimpleBi ''MsgStreamUpdate [
+    Cons 'MsgStreamUpdate [
+        Field [| msuWindow :: Word32 |]
+    ]]
+
+data MsgStream
+    = MsgStart MsgStreamStart
+    | MsgUpdate MsgStreamUpdate
+    deriving (Eq, Show, Generic)
+
+instance Bi MsgStream where
+    encode = \case
+        MsgStart s  -> encodeListLen 2 <> encode (0 :: Word8) <> encode s
+        MsgUpdate u -> encodeListLen 2 <> encode (1 :: Word8) <> encode u
+    decode = do
+        enforceSize "MsgStream" 2
+        tag <- decode @Word8
+        case tag of
+            0 -> MsgStart  <$> decode
+            1 -> MsgUpdate <$> decode
+            t -> cborError $ "MsgStream wrong tag: " <> show t
+
 data MsgStreamBlock
     = MsgStreamBlock Block
     | MsgStreamNoBlock Text
     | MsgStreamEnd
     deriving (Eq, Show, Generic)
+
+instance Bi MsgStreamBlock where
+    encode = \case
+        MsgStreamBlock b -> encodeListLen 2 <> encode (0 :: Word8) <> encode b
+        MsgStreamNoBlock t -> encodeListLen 2 <> encode (1 :: Word8) <> encode t
+        MsgStreamEnd -> encodeListLen 2 <> encode (2 :: Word8) <> encode (0 :: Word8)
+    decode = do
+        enforceSize "MsgBlock" 2
+        tag <- decode @Word8
+        case tag of
+            0 -> MsgStreamBlock <$> decode
+            1 -> MsgStreamNoBlock <$> decode
+            2 -> do
+                 (_ :: Word8 )<- decode
+                 pure MsgStreamEnd
+            t -> cborError $ "MsgStreamBlock wrong tag: " <> show t
```

Comment:
Orphans reunited with their type families is truly a heartwarming scene

Reply:
That's the kind of nice people we are around here :).

Reply:
I await `{-# LANGUAGE Illegalnstances #-}`...